python/triton/language/extra/hip/libdevice.py

```diff
@@ -0,0 +1,438 @@
+from triton.language import core
+
+
+@core.builtin
+@core._tensor_member_fn
+def abs(x, _builder=None):
+    x = core._to_tensor(x, _builder)
+    if x.dtype.is_floating():
+        return core.tensor(_builder.create_fabs(x.handle), x.type)
+    elif x.dtype.is_int_signed():
+        return core.tensor(_builder.create_iabs(x.handle), x.type)
+    else:
+        assert False, f"Unexpected dtype {x.dtype}"
+
+
+@core.builtin
+@core._tensor_member_fn
+def floor(x, _builder=None):
+    x = core._to_tensor(x, _builder)
+    if x.dtype.is_floating():
+      return core.tensor(_builder.create_floor(x.handle), x.type)
+    else:
+      assert False, f"Unexpected dtype {x.dtype}"
+
+
+@core.extern
+def rsqrt(arg0, _builder=None):
+    return core.extern_elementwise(
+        "", "", [arg0], {
+            (core.dtype("fp32"), ): ("__ocml_rsqrt_f32", core.dtype("fp32")),
+            (core.dtype("fp64"), ): ("__ocml_rsqrt_f64", core.dtype("fp64")),
+        }, is_pure=True, _builder=_builder)
+
+
+@core.extern
+def ceil(arg0, _builder=None):
+    return core.extern_elementwise(
+        "", "", [arg0], {
+            (core.dtype("fp64"), ): ("__ocml_ceil_f64", core.dtype("fp64")),
+            (core.dtype("fp32"), ): ("__ocml_ceil_f32", core.dtype("fp32")),
+        }, is_pure=True, _builder=_builder)
+
+
+@core.extern
+def trunc(arg0, _builder=None):
+    return core.extern_elementwise(
+        "", "", [arg0], {
+            (core.dtype("fp64"), ): ("__ocml_trunc_f64", core.dtype("fp64")),
+            (core.dtype("fp32"), ): ("__ocml_trunc_f32", core.dtype("fp32")),
+        }, is_pure=True, _builder=_builder)
+
+
+@core.extern
+def exp2(arg0, _builder=None):
+    return core.extern_elementwise(
+        "", "", [arg0], {
+            (core.dtype("fp32"), ): ("__ocml_exp2_f32", core.dtype("fp32")),
+            (core.dtype("fp64"), ): ("__ocml_exp2_f64", core.dtype("fp64")),
+        }, is_pure=True, _builder=_builder)
+
+
+@core.extern
+def sqrt(arg0, _builder=None):
+    return core.extern_elementwise(
+        "", "", [arg0], {
+            (core.dtype("fp32"), ): ("__ocml_sqrt_f32", core.dtype("fp32")),
+            (core.dtype("fp64"), ): ("__ocml_sqrt_f64", core.dtype("fp64")),
+        }, is_pure=True, _builder=_builder)
+
+
+# TODO: take from MLIR
+@core.extern
+def llrint(arg0, _builder=None):
+    return core.extern_elementwise(
+        "", "", [
+            arg0,
+        ], {
+            (core.dtype("fp32"), ): ("__nv_llrintf", core.dtype("int64")),
+            (core.dtype("fp64"), ): ("__nv_llrint", core.dtype("int64")),
```

Comment:
We have these functions in ocml lib. We can call them directly.

Reply:
And same for some other functions.

Reply:
It is interesting to see that `rint` is, in fact, results in intrinsic function

https://github.com/dfukalov/ROCm-Device-Libs/blob/45696ba432d766ec18295379555b9021299067fd/ocml/src/rintF.cl#L10-L14 

https://github.com/dfukalov/ROCm-Device-Libs/blob/45696ba432d766ec18295379555b9021299067fd/ocml/src/builtins.h#L155

Reply:
I wrote a small test HIP program
```cpp
// ex.cpp
template<typename T, typename D>
void __global__ kernel(T* iPtr, D* oPtr, unsigned size) {
  int i = hipBlockDim_x * hipBlockIdx_x + hipThreadIdx_x;
  T elem = iPtr[i];
  oPtr[i] = __builtin_rintf(elem);
}
template void __global__ kernel(float*, float*, unsigned);
```

Here is the way to compile it
```bash
/opt/rocm/llvm/bin/clang++ -DUSE_PROF_API=1 -D__HIP_PLATFORM_AMD__=1 -isystem /opt/rocm/include -std=c++17 --save-temps -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false -x hip ./ex1.cpp -emit-llvm -S -O3 -o ./ex1.ll
```

and here is the llvm dump

```bash
; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: readwrite)
define protected amdgpu_kernel void @_Z6kernelIffEvPT_PT0_j(ptr addrspace(1) nocapture readonly %0, ptr addrspace(1) nocapture writeonly %1, i32 %2) local_unnamed_addr #0 comdat {
  %4 = tail call ptr addrspace(4) @llvm.amdgcn.implicitarg.ptr()
  %5 = tail call i32 @llvm.amdgcn.workgroup.id.x()
  %6 = getelementptr inbounds i16, ptr addrspace(4) %4, i64 6
  %7 = load i16, ptr addrspace(4) %6, align 4, !tbaa !7
  %8 = zext i16 %7 to i64
  %9 = zext i32 %5 to i64
  %10 = mul nuw nsw i64 %8, %9
  %11 = tail call i32 @llvm.amdgcn.workitem.id.x(), !range !11, !noundef !12
  %12 = zext i32 %11 to i64
  %13 = add nuw nsw i64 %10, %12
  %sext = shl i64 %13, 32
  %14 = ashr exact i64 %sext, 32
  %15 = getelementptr inbounds float, ptr addrspace(1) %0, i64 %14
  %16 = load float, ptr addrspace(1) %15, align 4, !tbaa !13, !amdgpu.noclobber !12
  %17 = tail call contract float @llvm.rint.f32(float %16)
  %18 = getelementptr inbounds float, ptr addrspace(1) %1, i64 %14
  store float %17, ptr addrspace(1) %18, align 4, !tbaa !13
  ret void
}
```

So, in this particular case, we need to force `ExternElementwiseOp` to be re-written to `@llvm.rint.f32` intrinsic (see [here](https://mlir.llvm.org/docs/Dialects/LLVM/#llvmcall_intrinsic-llvmcallintrinsicop)).

Reply:
Hmm, `llvm.rint.f32` takes `f32` and returns `f32` (see [here](https://llvm.org/docs/LangRef.html#id643)). __nv_llrintf` returns `int64`